### PR TITLE
[MNT] Auto tag and label Docker images on release

### LINKED ISSUE #97: Dependabot cannot update template wf with variables?
Related to:
- https://github.com/neurobagel/workflows/issues/55

The template workflow https://github.com/neurobagel/workflows/blob/main/template_workflows/auto_release/build_docker_on_release.yml uses jinja-style variables in its body as placeholders for different repo names, as documented here: https://github.com/BetaHuhn/repo-file-sync-action?tab=readme-ov-file#using-templates.

The issue is that dependabot seems to have problems parsing this template .yml file (possibly because the variable syntax means the file is no longer technically valid YAML). 

Here's the result of manually triggering dependabot to check for updates:
image

As a result, the versions of GH actions used in `build_docker_on_release.yml` cannot be properly updated, meaning that whenever a change to these templates is synced to other repos, the updated versions of actions used in those repos is rolled back in the PR.

As an alternative, we may want to look into inheritable templates instead in https://github.com/BetaHuhn/repo-file-sync-action?tab=readme-ov-file#using-templates.

### LINKED ISSUE #100: Encode semver for `latest` image using Docker image label
If I pull the `latest` Docker images for my Neurobagel stack, I want to be able to double check what the actual numeric version is for the tools in case the `latest` version changes.

Because each tag for an image ends up in DockerHub as a separate pointer/entry, when you pull only `latest`, there's no straightforward way to out what other tags are associated with that specific digest from the command line. (Technically, the general recommendation for reproducibility is to use the semver tag directly.)

Since tags are not a viable way to encode version info about the `latest` image, we can use https://docs.docker.com/config/labels-custom-metadata/ Docker labels on the mutable `latest` tag instead to encode the corresponding semantic/persistent version.

`bids/validator` images for example do this, as shown in the config labels for the `latest` image:
```
            "Labels": {
                "org.opencontainers.image.created": "2024-05-09T21:19:41.620Z",
                "org.opencontainers.image.description": "Validator for the Brain Imaging Data Structure",
                "org.opencontainers.image.licenses": "MIT",
                "org.opencontainers.image.revision": "c790550f97b63c280a4e8c24c2c9805b7dbd8d5f",
                "org.opencontainers.image.source": "https://github.com/bids-standard/bids-validator",
                "org.opencontainers.image.title": "bids-validator",
                "org.opencontainers.image.url": "https://github.com/bids-standard/bids-validator",
                "org.opencontainers.image.version": "v1.14.6"
            }
```

(Can also grab labels directly with `docker image inspect --format='{{json .Config.Labels}}' myimage`).

### Steps to implement

- [x] Modify GH wfs for building & pushing Docker images on release to include a label with the semver, following https://docs.docker.com/build/ci/github-actions/manage-tags-labels/ 
- [x] Decide which metadata we want to include in the labels:
  - source / url
  - revision
  - version
- [ ] Document how to see what version a given `latest` image is assoc'd with

Nice to have: It would be nice if Docker could let a user know somehow if there is a more up-to-date `latest` version they could pull

### PR DESCRIPTION
**Pull request by alyssadai**
- Closes #100
- Fixes #97 

Changes proposed in this pull request:

- Use Docker Metadata Action to auto-tag images with semver `vX.Y.Z` and `latest`, and label them with the source code repo URL, revision ID, and semver
- Remove use of variable templating when syncing wf for pushing Docker images on release, which was problematic to keep for several reasons
    1. the template syntax doesn't work well with GH variable expansion in the same wf (see https://github.com/neurobagel/workflows/pull/49), meaning we wouldn't be able to use Docker Metadata Action without additional variable configurations in sync.yml
    2. we've already seen that this introduces problems for Dependabot parsing because it apparently technically allows what is invalid yaml syntax
- New repository variable `DOCKERHUB_REPO` has been added to repos that inherit this wf

## Checklist
_This section is for the PR reviewer_

- [x] PR has an interpretable title with a prefix (`[ENH]`, `[FIX]`, `[REF]`, `[TST]`, `[CI]`, `[MNT]`, `[INF]`, `[MODEL]`, `[DOC]`) _(see our [Contributing Guidelines](https://neurobagel.org/contributing/CONTRIBUTING#pull-request-guidelines) for more info)_
- [x] PR links to GitHub issue with mention `Closes #XXXX`
- [ ] Tests pass
- [x] Checks pass

For new features:
- [ ] Tests have been added

For bug fixes:
- [ ] There is at least one test that would fail under the original bug conditions.

### CODE REVIEW

File: .github/sync.yml [hunk 1]
```diff
@@ -40,6 +40,18 @@ group:
       neurobagel/tools
       neurobagel/recipes
 
+  # Workflow to build + push SEMVER Docker image on release
+  # NOTE: Each repo needs to have a secret DOCKERHUB_REPO set up which refers to the DockerHub repository name
```
Comment (surchs) [inline .github/sync.yml:44]:
```suggestion
  # NOTE: Each repo needs to have a secret with the name DOCKERHUB_REPO set up 
  # which refers to the DockerHub repository name
```

File: template_workflows/auto_release/build_docker_on_release.yml [hunk 2]
```diff
@@ -11,12 +11,33 @@ jobs:
       -
         name: Checkout
         uses: actions/checkout@v4
+      - 
+        # The release event is a superset of the push tag event,
```
Comment (surchs) [inline template_workflows/auto_release/build_docker_on_release.yml:15]: I'm not sure I understand what you mean here. When we make a release, we expect two tags to be made? Or every GH tag event is also a release event?

File: template_workflows/auto_release/build_docker_on_release.yml [hunk 3]
```diff
@@ -11,12 +11,33 @@ jobs:
...
+        # References:
+        # - https://github.com/docker/metadata-action?tab=readme-ov-file#typesemver
+        # - https://github.com/docker/metadata-action?tab=readme-ov-file#latest-tag
+        # - https://github.com/docker/metadata-action/issues/103
+        # - https://github.com/docker/metadata-action/issues/168
+        name: Docker meta
+        id: meta
+        uses: docker/metadata-action@v5
+        with:
+          # TODO: Set DOCKERHUB_REPO in the repository secrets (or in the variables)?
```
Comment (surchs) [inline template_workflows/auto_release/build_docker_on_release.yml:29]: variables is probably sufficient - the repo is public

File: template_workflows/auto_release/build_docker_on_release.yml [hunk 4]
```diff
@@ -27,4 +48,10 @@ jobs:
           context: .
           file: ./Dockerfile
           push: true
-          tags: {{ synced_repo.dockerhub_username }}/{{ synced_repo.docker_repo_name }}:latest, {{ synced_repo.dockerhub_username }}/{{ synced_repo.docker_repo_name }}:{{ synced_repo.release_tag }}
+          tags: ${{ steps.meta.outputs.tags }}
+          # Select a subset of labels to apply from https://github.com/opencontainers/image-spec/blob/main/annotations.md#pre-defined-annotation-keys,
+          # following https://github.com/docker/metadata-action/issues/303
```
Comment (surchs) [inline template_workflows/auto_release/build_docker_on_release.yml:53]: very cool

**Review by surchs — APPROVED** [review]
very cool, thanks @alyssadai. 

Was the build_on_release workflow the only one with the special sync_wf  double-braces syntax that was breaking dependabot? I.e. is the dependabot issue resolved because we no longer use the special double-braces syntax anywhere now? If so, let's make that clear in the issue as well so we don't have to guess later how we fixed it when the issue might come back.

:cook:

**Comment by alyssadai** [comment]
> Was the build_on_release workflow the only one with the special sync_wf double-braces syntax that was breaking dependabot? I.e. is the dependabot issue resolved because we no longer use the special double-braces syntax anywhere now?

Yes, that's right - we no longer use `{{ }}` for variable interpolation. I've added a comment to #97 for clarity.

**Pull request merged**